File: 1961-MaximumIceCreamBars/1961-MaximumIceCreamBars.cpp
// Last updated: 6/29/2026, 10:21:33 AM
class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        int maxCost = *max_element(costs.begin(), costs.end());

        vector<int> count(maxCost + 1, 0);
        for(auto &c : costs) {
            count[c]++;
        }

        int res = 0;

        for(int i = 1; i <= maxCost; i++) {
            if(count[i] == 0) continue;

            int canBuy = coins / i;
            if(canBuy <= count[i]) {
                return res + canBuy;
            }
            res += count[i];
            coins -= count[i] * i; 
        }
        return res;
    }
};

File: 1961-MaximumIceCreamBars/1961-MaximumIceCreamBars.py
# Last updated: 6/29/2026, 10:21:28 AM
class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        maxCost = max(costs)

        count = [0] * (maxCost + 1)

        for c in costs:
            count[c] += 1
        
        res = 0
        for i in range(1, maxCost + 1):
            if count[i] == 0:
                continue
            canbuy = coins // i

            if(canbuy <= count[i]):
                return res + canbuy
            
            res += count[i]
            coins -= count[i] * i
        
        return res
